### DISCUSSION TITLE
Is Avalonia install causing issues with my non-Avalonia dotnet framework apps?

### DISCUSSION THREAD

**Discussion by lansalot** [AvaloniaUI/Avalonia#14277] [Q&A]
Afternoon all

There's a post on this here: [Error CS8630 Invalid 'nullable' value: 'Enable' for C# 7.1. Please use language version '8.0' or greater](https://stackoverflow.com/questions/77845971/all-vs-projects-now-throwing-errorcs8630invalid-nullable-value-enable-for)

I today noticed that any and all of my existing winforms stuff that targets 4.8.1 framework are throwing the above error and don't compile. There's a workaround I found while looking into it, and one of them references Avalonia in the project the guy is having trouble with. That's the only thing I've done recently that I can think of - install the Avalonia framework.

Wondered if it could be related, and what I can do as I am taking part in multiple non-Avalonia related projects and don't want to be checking the workaround into code that's used globally. I'd like to keep Avalonia installed for newer stuff I'm looking into, so removing it not the preferred option.

Thanks
A

**Comment by thevortexcloud** [comment]
Nullable reference annotations is typically enabled by a project setting. It should not be affected by dependencies. It's enabled by default for new projects so if you have upgraded your project file recently from the old format it's probably been enabled.

You can also force disable it using a preprocessor.

https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/preprocessor-directives

That error can also happen if you try to use one of the features of null reference annotations.

**Reply by lansalot** [reply]
Thanks, I've seen similar while searching for the issue - but again, these are existing projects where I haven't changed a thing (git history proves this). Projects I could download a zip for and compile are now throwing this error, and I've no clue why - Avalonia installation is the only thing I can think of that I did recently. I tried uninstalling it, but it's made no difference.

Not saying it is Avalonia for sure of course, but for nothing except windows updates, it seems very strange?

What global setting might cause this, something in VS options perhaps?

**Reply by thevortexcloud** [reply]
Visual Studio could just be doing weird things. It does that _a lot_ (I have personally wasted several hours on Visual Studio insisting things are broken when they were not). Have you:

* Confirmed it actually builds using the CLI (if you are using the new project format, you just need to run `dotnet build`)?
* Tried doing a clean and a rebuild? 
* Tried restarting visual studio?
* Tried doing a clean, closing visual studio, reopening visual studio, then doing a build?
* Checked nobody has accidentally made a typo somewhere in a previous commit? EG declaring a nullable reference type such as `string?` or using a `!` to ignore a nullable value.

But I am not aware of any setting in Visual Studio to force enable it. As the error suggests, it's a language feature that needs to be opted into. It would be weird if that was a global setting.

But seriously, half the time Visual Studio raises things as errors they are ignorable or fixable by rebooting Visual Studio.

**Reply by lansalot** [reply]
Same in CLI, cleaned/rebuilt, restarting, rebooting, you name it. Again, all code that worked fine recently. Project also builds fine in github actions.

Just opened a solution, untouched since 2022. Same error for all 3 projects.

My tablet doesn't have Avalonia on it, so I tried compiling the exact same project on it. Same error !

So, not an Avalonia issue, and not an issue particular to one machine (tablet is on 17.8.4, so pre-VSupdate). Both machines have had the Jan24 updates tho...

**Comment by lansalot** [comment]
Following guidance here:
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/configure-language-version

The code was saying 7.1, I changed it to 7.3 expecting the same error yep, got the error. Change it to 8.0 and it worked fine. 
![image](https://github.com/AvaloniaUI/Avalonia/assets/9885921/f068bed8-621f-4071-bbb4-2dff3bdd2004)

In the main repository, it says 7.1, and if I download that and run it.. it works fine!

Anyway, not an Avalonia issue for sure, so I'll close this one off I guess - unless anyone has any further suggestions?

**Comment by lansalot** [comment]
So, on Stackoverflow, someone asked if I had a Directory.Build.Props file anywhere?

Sure enough, in the root of my github folder, there was one - with this:

```
<Project>
  <PropertyGroup>
    <Nullable>enable</Nullable>
    <AvaloniaVersion>11.0.2</AvaloniaVersion>
  </PropertyGroup>
</Project>
```

So I guess that rippled down into all my non-Avalonia projects. Renamed the file, restarted VS, all projects fine again. What a nightmare. Was this some bad option I picked during install/config perhaps?

I'm suspecting perhaps I did the dotnet cmd from wrong subdir..?
`dotnet new install Avalonia.Templates`

**Reply by stevemonaco** [reply]
This seems to be added when you use the xplat template (and not the others). I don't think there's an expectation of having legacy stuff side-by-side with Avalonia from the template's perspective. The props file comes from [this in avalonia-dotnet-templates](https://github.com/AvaloniaUI/avalonia-dotnet-templates/blob/master/templates/csharp/xplat/Directory.Build.props) and you should create an issue in that repo if you feel that it warrants a change. It's a valid complaint to be considered and the template can make these things optional.

There's probably some expectation that devs understand newer .NET build concepts though. Managing `.props` files, updating C# lang versions (and polyfill via [PolySharp](https://github.com/Sergio0694/PolySharp)), or setting `<Nullable>disable</Nullable>` in the legacy projects are viable options.
